src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,215 @@ namespace mlpack {
-  arma::uvec indices = randi<arma::uvec>(dataset.n_cols,
+    // Random sampling with replacement.
+    arma::uvec indices = randi<arma::uvec>(dataset.n_cols,
       arma::distr_param(0, dataset.n_cols - 1));
-  bootstrapDataset = dataset.cols(indices);
-  bootstrapLabels = labels.cols(indices);
-  if (UseWeights)
-    bootstrapWeights = weights.cols(indices);
-}
+    bootstrapDataset = dataset.cols(indices);
+    bootstrapLabels = labels.cols(indices);
+    if (UseWeights)
+      bootstrapWeights = weights.cols(indices);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ *
+ * @tparam IndMatType Indicator matrix type.
+ */
+template<typename IndMatType = arma::umat>
+class SequentialBootstrap
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param[in] intervals See ComputeAverageUniqueness().
+   */
+  SequentialBootstrap(const IndMatType& intervals) :
+    intervals(intervals)
+  {
+    if (intervals.n_cols != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2 x m matrix!");
+  }
+
+  template<bool UseWeights,
+    typename MatType,
+    typename LabelsType,
+    typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+    const LabelsType& labels,
+    const WeightsType& weights,
+    MatType& bootstrapDataset,
+    LabelsType& bootstrapLabels,
+    WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != intervals.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "labels n_cols and intervals n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+    const arma::uvec phi(arma::conv_to<arma::uvec>::from(
+      ComputeSamples(dataset.n_cols)));
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  /**
+   * Compute the samples of the next draw.
+   *
+   * @param[in] colCount Number of data points in the dataset.
+   *
+   * @return A list of indices referring to the observations that should
+   *         be sampled.
+   */
+  std::vector<arma::uword> ComputeSamples(arma::uword colCount) const
```

Comment:
I think it would be better to just return an `arma::uvec` here.  You can instantiate it to the size `colCount` at the beginning of the method, and then modify `ComputeNextDrawProbabilities` to also take `i`, which represents the number of elements of `phi` that are currently used.  That will also simplify a lot of the other conversion code that you are doing later on.